augur/frequencies.py

```diff
@@ -41,72 +64,121 @@ def get_pivots(tps, dt,min_date, max_date):
+
+                tree_freqs = tree_frequencies(tree, pivots, method='SLSQP',
+                                              node_filter = node_filter_func,
+                                              ws = max(2, tree.count_terminals()//10),
+                                              stiffness = stiffness)
+
+                tree_freqs.estimate_clade_frequencies()
+
+                if args.output_format == "nextflu":
+                    # Export frequencies in nextflu-format by region and clade id.
+                    for clade_id, clade_frequencies in tree_freqs.frequencies.items():
+                        frequency_dict["%s_clade:%d" % (region, clade_id)] = format_frequencies(clade_frequencies)
+
+                    frequency_dict["%s:counts"%region] = [int(x) for x in tree_freqs.counts]
+                else:
+                    # Export frequencies in auspice-format by strain name.
+                    for node in tree.find_clades():
+                        if node.is_terminal() or args.include_internal_nodes:
+                            frequency_dict[node.name] = {
+                                "frequencies": format_frequencies(tree_freqs.frequencies[node.clade])
+                            }
+        elif args.method == "kde":
+            if args.output_format == "nextflu":
+                print("ERROR: nextflu format is not supported for KDE frequencies", file=sys.stderr)
+                return 1
 
-            tree_freqs.estimate_clade_frequencies()
-            for x,val in tree_freqs.frequencies.items():
-                frequency_dict["%s_clade:%d"%(region,x)] = format_frequencies(val)
+            # Estimate frequencies by KDE method.
+            # Load weights if they have been provided.
+            if args.weights:
+                with open(args.weights, "r") as fh:
+                    weights = json.load(fh)
 
-            frequency_dict["%s:counts"%region] = [int(x) for x in tree_freqs.counts]
+                weights_attribute = args.weights_attribute
+            else:
+                weights = None
+                weights_attribute = None
+
+            # Estimate frequencies.
+            kde_frequencies = TreeKdeFrequencies(
+                sigma_narrow=args.narrow_bandwidth,
+                sigma_wide=args.wide_bandwidth,
+                proportion_wide=args.proportion_wide,
+                pivot_frequency=int(12.0 / args.pivots_per_year),
+                start_date=args.min_date,
+                end_date=args.max_date,
+                weights=weights,
+                weights_attribute=weights_attribute,
+                include_internal_nodes=args.include_internal_nodes,
+                censored=args.censored
+            )
+            frequencies = kde_frequencies.estimate(tree)
+
+            # Export frequencies in auspice-format by strain name.
+            frequency_dict = {"pivots": list(kde_frequencies.pivots)}
+            for node_name in frequencies:
```

Comment:
this should be de-duplicated with the equivalent block for diffusion frequencies

Reply:
I couldn't completely de-duplicate this code without some more reworking of the different frequency estimator classes. More details are in the commit message for 1dc8f03.